add: ledger local setup guide (#3160)
dfinity/portal: docs/developer-docs/defi/tokens/ledger/usage/icrc1_ledger_usage.mdx

---
keywords: [intermediate, tokens, icrc, tutorial, ledger]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

# Using the ICRC-1 ledger

<MarkdownChipRow labels={["Intermediate" , "Tutorial"]} />

## Overview

> There are two main ways to interact with an ICRC-1 ledger.
> - `dfx canister`: the generic canister call from `dfx`.
> - `ic-cdk`: inter-canister calls for the ICRC-1 ledger.
> - [`ledger-icrc-js`](https://github.com/dfinity/ic-js/tree/main/packages/ledger-icrc): a library for interfacing with ICRC-1 ledger on the Internet Computer.

Reviewer: ```suggestion
There are two main ways to interact with an ICRC-1 ledger.
- `dfx canister`: The generic canister call from `dfx`.
- `ic-cdk`: Inter-canister calls for the ICRC-1 ledger.
- [`ledger-icrc-js`](https://github.com/dfinity/ic-js/tree/main/packages/ledger-icrc): A library for interfacing with ICRC-1 ledger on the Internet Computer.
```

## ICRC-1 and ICRC-1 extension endpoints

> Whether your ICRC-1 ledger will have all the endpoints discussed in this tutorial will depend on whether you support any of the extensions of ICRC-1 (ICRC-2, ICRC-3,...).

Reviewer: ```suggestion
Whether your ICRC-1 ledger will have all the endpoints discussed in this tutorial will depend on if you chose to support any of the extensions of ICRC-1 (ICRC-2, ICRC-3,...).
```

> This tutorial will go through the endpoints for ICRC-1 and the extension ICRC-2. If you have deployed an ICRC-1 ledger according to the [guide](/docs/developer-docs/defi/tokens/ledger/setup/icrc1_ledger_setup.mdx) you have to make sure that you have enabled ICRC-2 standard. Otherwise these calls will not work.

Reviewer: ```suggestion
This tutorial will go through the endpoints for ICRC-1 and the extension ICRC-2. If you have deployed an ICRC-1 ledger according to the [guide](/docs/developer-docs/defi/tokens/ledger/setup/icrc1_ledger_setup.mdx) you have to make sure that you have enabled the ICRC-2 standard, otherwise, the calls demonstrated in this guide will not work.
```

```
(record { allowance = 100_000 : nat; expires_at = null })
```
Alternatively, you can also set the expiration date for approval. For specifics, see the [ICRC-2 standard](https://github.com/dfinity/ICRC-1/tree/main/standards/ICRC-2#icrc2_approve).

If the spender now wants to transfer tokens that were previously approved for the spender to a certain principal (in this case, use the arbitrary principal `7tmcj-ukheu-y6dvi-fhmxv-7qs5t-lwgh2-qzojr-vzt6m-maqv4-hvzlg-5qe`) you can call:
```
dfx canister call icrc1_ledger_canister icrc2_transfer_from "(record { amount = 90_000; from = record{owner = principal \"${DEFAULT}\"}; to= record{owner = principal \"${DEFAULT}\"}; })"
```
Note that you cannot transfer the entire allowance as the fee for making a transfer has to be substracted from the transferred amount. For the reference implementation the fee is `10_000` tokens. Thus, the maximum amount that can be transferred from `DEFAULT` to `7tmcj-ukheu-y6dvi-fhmxv-7qs5t-lwgh2-qzojr-vzt6m-maqv4-hvzlg-5qe` with the spender `sckqo-e2vyl-4rqqu-5g4wf-pqskh-iynjm-46ixm-awluw-ucnqa-4sl6j-mqe` is `90_000`.

> The principal making this call has to be the spender, since they are the once that received the approval.

Reviewer: ```suggestion
The principal making this call has to be the spender, since they are the one that received the approval.
```

You will receive the block index as a return value:
```
(variant { Ok = 3 : nat })
```

## Interacting with an ICRC-1 ledger from another canister (inter-canister calls via `ic-cdk`)
You can look at the documentation of [inter-canister calls](/docs/current/developer-docs/backend/rust/intercanister) to see how you can interact with another canister from inside a canister. This guide will give you a couple of examples of how to make such a call in the case of the ICRC-1 ledger.

Here is an example of how to fetch the name from the ICRC-1 ledger using Rust and the `ic-cdk` [library](https://github.com/dfinity/cdk-rs) from within a canister:

> You will need the principal of the ICRC-1 ledger. For this guide,  take the canister ID that was used in the previous guide on [deploying an ICRC-1 ledger](/docs/developer-docs/defi/tokens/ledger/setup/icrc1_ledger_setup.mdx), which was `mxzaz-hqaaa-aaaar-qaada-cai`.

Reviewer: ```suggestion
You will need the principal of the ICRC-1 ledger. For this guide, take the canister ID that was used in the previous guide on [deploying an ICRC-1 ledger](/docs/developer-docs/defi/tokens/ledger/setup/icrc1_ledger_setup.mdx), which was `mxzaz-hqaaa-aaaar-qaada-cai`.
```
